### ISSUE TITLE
Port "capture" interface into this library from main spectral-workbench project

### ISSUE THREAD

**Issue by jywarren** [publiclab/spectral-workbench.js#56]
The live Capture interface is a great thing to bring into here -- https://spectralworkbench.org/capture

https://github.com/publiclab/spectral-workbench/blob/main/app/views/capture/_capture.html.erb and other files in that directory are the HTML portion.

The JavaScript is in https://github.com/publiclab/spectral-workbench/blob/main/app/assets/javascripts/capture.js

The first attempt, we can just copy that whole file over, and have it appear locally in `/src/`. 

We can consolidate the HTML files into a single static file in https://github.com/publiclab/spectral-workbench.js/tree/main/examples/capture.html and make sure it's pointed at capture.js, and add any necessary dependencies to our package.json file, installing them via `npm`. 

Once we get it all running, we can start doing some awesome things like getting it running on a Raspberry Pi! See this code for reference: https://github.com/publiclab/infragram/blob/f5f62ce483fbc88823de96f5c87d160228c0a764/pi/index.html#L189-L191

We'd love help with this one! A lot of moving parts, but a fun one :-)

**Comment by jywarren** [comment]
I think this would do to get a Pi running:

```
var piImage = new Image();
$W.getCrossSection = function() {
  $W.ctx.drawImage(piImage, 0, -$W.sample_start_row);
  piImage.src = "http://pi.local/cam/cam_pic.php?time=" + new Date().getTime(); 
}
```

This could be seriously improved by getting the preview window working. But it'll do OK for starters. I think the images could be drawn to the `<video>` element too... but all this is a follow-up issue, don't worry about this just yet.

**Comment by jywarren** [comment]
OK, I've done a very minimal initial implementation here: https://github.com/publiclab/spectral-workbench.js/pull/57

This needs a good bit of work, but is successfully getting video from the webcam and graphing.

**Comment by jywarren** [comment]
It still needs debugging, as well as better separation of basic UI from the integration with the parent server Rails code. Ideally the specific interface connections would be set up via a more standardized set of HTML elements with unique IDs, which would be bound to abstracted features of the capture code. Right now, it's all munged together!

**Comment by jywarren** [comment]
But an initial step could be to comment out server-specific code and connect it up for us on the Raspberry Pi, and stand-alone in github pages.

**Comment by jywarren** [comment]
We'll also want this code to produce data in the formats described in https://github.com/publiclab/spectral-workbench.js#usage

**Comment by jywarren** [comment]
Hi @dhiraj240! Thanks for your interest! We don't really reserve projects, but you are very welcome to indicate this as the subject of your GSoC proposal, and that would be great!

I think there are some proposals for neural network or machine learning models, and i think there is some potential for this, but I think perhaps it makes sense for such ideas to be implemented in parallel as swap-in alternatives for existing features, such as the linear calibration we currently use. That way people could mix and match, and the substantial code that might go into such a model could be isolated from the main codebase in a modular way. But I like the idea - as do I like the idea of using machine learning to match spectra, or identify features, perhaps. I'd also recommend opening a separate issue for this and linking in related ideas from https://github.com/publiclab/spectral-workbench/issues/399, https://github.com/publiclab/plots2/issues/4660, and potentially some of the ideas from https://publiclab.org/notes/warren/01-02-2019/brainstorming-for-summer-of-code-2019 (where i recommend you leave a comment about your ideas if you haven't already!)

For the camera integration, yes, I've done a crude initial move of the code, but SpectralWorkbench.org is not running this capture code yet, and it'd require some cleanup and integration work to get it to be included over there. Worthwhile! And the most related to this issue topic. 

For the Ruby server, I don't think we're likely to port the entire system, to tell the truth; the code is substantial enough that I would prefer to think about ways to move some more of the functionality into client-side javascript, while keeping the data storage in the back-end Ruby system. One thing that could be an important step towards shifting Ruby code into JS, however, would be to formalize, document, and write tests for the API that is required to provide data to the client-side code. Right now it's spread among a set of different calls that aren't very consistent and most do not have tests. Opening a separate issue for this kind of project, like `standardize, document, and write tests for JavaScript API` on [the parent Ruby project](https://github.com/publiclab/spectral-workbench/issues/) would be very welcome, and might also enable alternative clients to connect through the resulting standard APIs. 

Thanks for your great energy!

**Comment by jywarren** [comment]
Also interesting! https://github.com/publiclab/spectral-workbench.js/issues/75

**Comment by Dhiraj240** [comment]
Hmm. Things look more clearer to me.But right now i am confused as from where should i start ?
Should i be working on connecting raspberry pi as suggested above.

**Comment by jywarren** [comment]
I think the best course would be to start refactoring the capture code to
follow a structure similar to the code in, say, publuclab/image-sequencer -
something that gets built with browserify, and has tests, so we can get the
under the surface stuff better organized and easier to read and maintain,
then work on the integration parts. Does this make sense?

**Comment by Dhiraj240** [comment]
Okay, things seem clearer now !! i will start working over it.I had a long gap since i joined this organisation and now i got something to prove.
Thank You Sir !! :smile:

**Comment by jywarren** [comment]
Thank you too, and let's work on this in small pieces please, so that we
can collaborate! There are others who join our weekly check ins that have
done architectural work on breaking up and restructuring libraries, most
recently in image sequencer. Say hello in the check in and see if others
can offer some help!

**Comment by Dhiraj240** [comment]
@jywarren  I have started refracting the ```capture.js``` code file. I will submit my PR by tomorrow, should i send the PR without getting assigned.
If not then assign this issue and #75  to me. After a hard time i configured my Pi. :smile:

**Comment by jywarren** [comment]
oh good to hear it! Great!!! Good luck and looking forward to it!!!

**Comment by sidntrivedi012** [comment]
@jywarren Have you completed the porting of the javascript part of SWB in #57 ?

**Comment by jywarren** [comment]
No, i haven't!

**Comment by sidntrivedi012** [comment]
Ok, just wanted to confirm. Thanks :slightly_smiling_face:

**Comment by jywarren** [comment]
Hi @sidntrivedi012 just copying in helpful links from the Open Call!


spectralworkbench.org - spectralworkbench.org/capture 
- making a spectrometer to test : https://publiclab.org/papercraft-spec 
- Complete the refactoring of capture part into spectral-workbench.js
- https://github.com/publiclab/spectral-workbench.js/pull/57 
- https://github.com/publiclab/spectral-workbench.js/tree/main/examples/capture 
- https://publiclab.github.io/spectral-workbench.js/examples/capture/ 
- https://github.com/publiclab/spectral-workbench.js/blob/main/examples/capture/capture.js
- Standardized library for webcam API: https://github.com/publiclab/spectral-workbench.js/issues/87

**Comment by jywarren** [comment]
I believe this is largely complete in https://github.com/publiclab/spectral-workbench.js/pull/175 !!!
